### ISSUE TITLE
How set Cookie and Session Timeout Based on Logged In User

### ISSUE THREAD

**Issue by mschaefer-gresham** [dotnet/aspnetcore#54966]
I'm setting the cookie and session timeout as follows (I need to use this API. I can't use the new Identity api. Please don't provide solutions based on Identity):

```
.AddCookie(options =>
{
    options.ExpireTimeSpan = TimeSpan.FromMinutes(15);
    options.SlidingExpiration = true;
    options.Cookie.MaxAge = options.ExpireTimeSpan;
})
.AddOpenIdConnect(options =>
{
     options.MaxAge = TimeSpan.FromMinutes(15);
     ...
});
```

This sets the same timeout for every user. And these values are set before the user is redirected to the IDP.

How can I update both the cookie and session timeout to different values once the user has logged in (but before the cookie is returned to the UI). I'm authenticating against Azure B2C and would like to use the user's role to determine the timeout.

I'm hoping there is an event I can listen for that is passed information returned from Azure that would allow me to update the timeouts once the user has successfully logged in.

**Comment by mschaefer-gresham** [comment]
@Raajdee-P thanks for the response, but this doesn't not solve the issue. There is absolutely no difference from what I have. I'm also setting Authority, ClientId, and ClientSecret (I left them out for clarity).

**Comment by halter73** [comment]
> How can I update both the cookie and session timeout to different values once the user has logged in (but before the cookie is returned to the UI). I'm authenticating against Azure B2C and would like to use the user's role to determine the timeout.

You haven't shown any code related session, but that's normally not tied to users and the timeouts can only be configured through the `IdleTimeout`, `IOTimeout` and `Cookie.MaxAge` properties on `SessionOptions`. While the session feature itself is completely oblivious to who the currently logged-in user, it's easy to effectively tie a session to a user by using the user id as part of the session key. If the user isn't logged in, they get an anonymous session.

You could try something like the follwoing to set the cookie expiration after you get the `ClaimsPrincipal` from Azure B2C.

```csharp
.AddOpenIdConnect(options =>
{
    options.Events.OnTicketReceived = context =>
    {
        if (context.Principal.IsInRole("Sensitive"))
        {
            context.Properties.ExpiresUtc = DateTimeOffset.UtcNow.AddMinutes(15);
        }
        else
        {
            context.Properties.ExpiresUtc = DateTimeOffset.UtcNow.AddDays(7);
        }
    };
});
```

Depending on your requirements, you may want to set `CookieAuthenticationOptions.SlidingExpiration` to `false`, so the cookie really does expire at the time you initially set. But if you're okay with keeping the user logged in as long as they're actively using the site, sliding expiration should work fine.

If you initially set expiration to 15 minutes in the future, the cookie will refresh if the server receives a request at any time from 7.5 to 15 minutes from now that will extend the expiration by another 15 minutes. If you initially set expiration to 7 days in the future, the cookie will only get refreshed after 3.5 days, but that refresh will extend the expiration by another 7 days. Either way, if a user continuously uses the site, they will never be forced to reauthenticate.

**Comment by mschaefer-gresham** [comment]
@halter73 thank you. I ended up solving this by setting the OpenIdConnect property UseTokenLifetime to true and removing the max age cookie setting. The token timeout is set in the B2C policy. I removed the max age OpenIdConnect setting as well from the init settings. Instead, I set it in options.Events.OnTokenValidated, however this may be redundant. And finally, I have to set the session IdleTimeout to a value that will be sufficient for all possible policies as the site supports multiple IDPs. Perhaps there is a better way, but this is working.
